velox/functions/sparksql/aggregates/FirstLastAggregate.cpp

```diff
@@ -152,18 +196,23 @@ class FirstAggregate : public FirstLastAggregateBase<numeric, TData> {
 
   // If we found a valid value, set to accumulator, then skip remaining rows in
   // group.
-  bool updateValue(vector_size_t i, char* group, DecodedVector& decoded) {
+  bool updateValue(
+      vector_size_t i,
+      char* group,
+      const DecodedVector& decodedVector,
+      const BaseVector* valueVector) {
     auto accumulator = exec::Aggregate::value<TAccumulator>(group);
     if (accumulator->has_value()) {
       return false;
     }
 
     if constexpr (!numeric) {
-      return updateNonNumeric(i, group, decoded);
+      return updateNonNumeric(i, group, decodedVector, valueVector);
     } else {
-      if (!decoded.isNullAt(i)) {
+      if (!decodedVector.isNullAt(i)) {
         exec::Aggregate::clearNull(group);
-        auto value = decoded.valueAt<TData>(i);
+        auto value = valueVector->as<SimpleVector<TData>>()->valueAt(
```

Comment:
Can you move this dynamic cast (valueVector->as<SimpleVector<TData>>()) outside of the loop to avoid doing it again and again for each row?